src/runtime.rs

```diff
@@ -153,7 +149,453 @@ pub struct FsSearchBody {
+                // However, it goes deeper than this, as some files are shared even further. For example,
+                // "fighter/roy/model/body/c00/model.xmb" redirects to "fighter/roy/model/body/c07/model.xmb"
+                // which further redirects to fox's model.xmb.
+                // In order to unshare these files we have to go two redirections deep to be 100% sure we map the right hashes together
+                let mut index_to_data_hash_map = HashMap::new();
+                for info in shared_load_data_infos.iter() {
+                    let hash = file_paths[usize::from(info.file_path_index)].path.hash40();
+                    let idx = file_info_indices[usize::from(info.file_info_indice_index)].file_info_index;
+                    let idx = file_infos[usize::from(idx)].file_path_index;
+                    index_to_data_hash_map.insert(idx, hash);
+                }
+                // we do this to be able to go from shared data hash -> mass load group hash
+                // this is important so that the file path indices in our new MassLoadData point
+                // to the original filepaths
+                let mut data_to_group_hash_map = HashMap::new();
+                let mut group_to_index_hash_map = HashMap::new();
+                for (offset, info) in mass_load_group_infos.iter().enumerate() {
+                    let idx = file_info_indices[usize::from(info.file_info_indice_index)].file_info_index;
+                    let idx = file_infos[usize::from(idx)].file_path_index;
+                    if let Some(data_hash) = index_to_data_hash_map.get(&idx) {
+                        let group_hash = file_paths[usize::from(info.file_path_index)].path.hash40();
+                        if *data_hash == group_hash {
+                            return Err(LookupError::Missing); // TODO: Change this to a better error code or smthn
+                        }
+                        data_to_group_hash_map.insert(*data_hash, group_hash);
+                        group_to_index_hash_map.insert(group_hash, offset);
+                    }
+                }
+                drop(index_to_data_hash_map);
+
+                // since we also have to include our redirection chain, we have to make sure we get contigious FileInfos
+                // for our new MassLoadData
+                let current_mld_offset = new_infos.len();
+                new_infos.resize_with(current_mld_offset + shared_load_data_infos.len(), || default_info);
+                for (offset, info) in shared_load_data_infos.iter().enumerate() {
+                    // get the shared hash
+                    let source_info_hash = file_paths[usize::from(info.file_path_index)].path.hash40();
+                    let new_info_hash = data_to_group_hash_map.get(&source_info_hash).expect(&format!("Could not find new hash for source {:#x?}", source_info_hash));
```

Comment:
these formats will run every time this line of code runs, even on success. this is not cheap.
I recommend something like:

```rust
let new_info_hash = match data_to_group_hash_map.get(&source_info_hash) {
    Ok(x) => x,
    Err(_) => panic!("Could not find new hash for source {:#x?}", source_info_hash)
};
```